dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/event/data/AbstractJdbcEventAnalyticsManager.java

```diff
@@ -821,53 +827,63 @@ else if ( header.getValueType() == ValueType.REFERENCE )
      */
     protected String getItemsSql( EventQueryParams params, SqlHelper hlp )
     {
-        List<String> repeatableStagesSqlList = new ArrayList<>();
+        Map<Boolean, List<QueryItem>> itemsByRepeatableFlag = params.getItems()
+            .stream()
+            .filter( QueryItem::hasFilter )
+            .collect( Collectors.groupingBy(
+                QueryItem::hasRepeatableStageParams ) );
+
+        Collection<String> repeatableSqlConditions = asSqlCollection( itemsByRepeatableFlag.get( true ), params );
+        Collection<String> nonRepeatableSqlConditions = asSqlCollection( itemsByRepeatableFlag.get( false ), params );
+
+        return (!repeatableSqlConditions.isEmpty() || !nonRepeatableSqlConditions.isEmpty() ? hlp.whereAnd() + " "
+            : "") +
+            Stream.of(
+                joinSql( repeatableSqlConditions, joining( " or ", "(", ")" ) ),
```

Comment:
I am wondering if this code can be written in a way which is a bit more readable. I had a hard time understanding what is going on here.

Reply:
Yes you're right.
Basically I need " and " when at least one of the two collections is not empty, then I need to join those collections using " OR " and " AND " respectively. Streams is because if collections are empty I don't need neither "(" and ")". I'll refactor it in a more readable way